File: Data-Structures-and-Algorithms/0141-LinkedListCycle/hasCycle.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution(object):
    def hasCycle(self, head):
        # use extra N space to save hashset of seen values
        nodes_seen = set()
        # when head is None, we know we are at the end of the linked list, and we stop the loop
        while head is not None:
            # return true for the node has been reached before
            if head in nodes_seen:
                return True
            # if it is a new node, then add to the set
            nodes_seen.add(head)
            print(head.val)
            # move to the next node till it is null 
            head = head.next
        return False

LL = ListNode(0)
LL.next = ListNode(1)
LL.next.next = ListNode(2)
LL.next.next.next = ListNode(3)
LL.next.next.next.next = ListNode(4)
LL.next.next.next.next.next = ListNode(5)
print(Solution().hasCycle(LL))

# O(N) one pass traverse
# O(N) constant extra space required
# Is it a cycle?

File: Data-Structures-and-Algorithms/0141-LinkedListCycle/hasCycle.js
function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
}

const hasCycle = function(head) {
    // if the head is empty, we immediately return False
    if ( head == null ) {
        return false;
    };

    // use two runners, one fast and one slow
    var slow = head;
    var fast = head.next;
    // if by the time when fast catches up to slow in the next cycle, we return true and stop the loop
    while ( slow != fast ) {
        // if we find a node that is null, we know it is not a cycle
        if ( fast == null || fast.next == null) {
            return false;
        }
        // update the runners 
        slow = slow.next;
        fast = fast.next.next;
    };
    
    return true
};

let LL = new ListNode(0)
LL.next = new ListNode(1)
LL.next.next = new ListNode(2)
LL.next.next.next = new ListNode(3)
LL.next.next.next.next = new ListNode(4)
LL.next.next.next.next.next = new ListNode(5)
LL.next.next.next.next.next.next = LL
console.log(hasCycle(LL))

// O(N)
// O(1) n
// Is it a cycle?